asyncio_run_in_process/_child.py

```diff
@@ -70,11 +72,69 @@ def update_state_finished(to_parent: BinaryIO, finished_payload: bytes) -> None:
     to_parent.flush()
 
 
+async def _do_task_cleanup(*tasks: 'asyncio.Future[Any]') -> None:
+    for task in tasks:
+        if not task.done():
+            task.cancel()
+
+        try:
+            await task
+        except asyncio.CancelledError:
+            pass
+
+
 SHUTDOWN_SIGNALS = {signal.SIGTERM}
 
 
+async def _handle_SIGTERM(task: 'asyncio.Task[TReturn]', fut: 'asyncio.Future[int]') -> None:
+    signum = await fut
+    raise SystemExit(signum)
+
+
+async def _handle_coro(coro: Coroutine[Any, Any, TReturn], got_SIGINT: asyncio.Event) -> TReturn:
+    coro_task = asyncio.ensure_future(coro)
```

Comment:
If this line is changed to:

```
coro_task = asyncio.ensure_future(asyncio.shield(coro))
```

then there isn't a need to catch the `RuntimeError` below, because cancelling the task will no longer attempt to cancel the coroutine we manually finished.

Reply:
I don't think that's actually the problem.  I believe the `RuntimeError` comes from the fact that by including `coro` in the call to `asyncio.wait(...)` it get's *awaited*.  Then, later in the case where `coro.throw(..)` gets called we're actually performing a semi-against-the-rules action since the `Task` should be the one in charge of managing the lifecycle of the coroutine.  But in order to inject the exception this was the only route I could find, so I believe the `RuntimeError` is inevitable but likely acceptable.

Reply:
Oh, I might have gotten the reason wrong but I know that wrapping it in `asyncio.shield` obviates the need for catching the `RuntimeError` :) And I think that's preferable; looking at the asyncio code there are a lot of reasons why `RuntimeError` might be thrown here, I don't think it's safe to ignore all of them!

My best-guess though is that when the `StopIteration` is thrown it means the generator has finished and cannot be resumed. Then, when `coro_task` is awaited during the call to `_do_task_cleanup`, the task [tries to run the generator](https://github.com/python/cpython/blob/master/Lib/asyncio/tasks.py#L280), but [that fails because the generator is already finished](https://github.com/python/cpython/blob/master/Objects/genobject.c#L178).

Reply:
Looking at [`asyncio.shield`](https://docs.python.org/3/library/asyncio-task.html#asyncio.shield) I don't see anything about `RuntimeError`.

As for blanket catching of `RuntimeError` being dangerous, :+1: I'll add some matching against the error text which is the best I can think of.

I agree with your explanation for *why* the `RuntimeError` shows up, but it doesn't alleviate the issue which is outlined in the new docstring:

We have a task that has not been awaited that is wrapping a coroutine which *if* awaited will raise a `RuntimeError`, but if not awaited will emit a warning.